### PR TITLE
enforce the active_connection_id_limit

### LINKED ISSUE #3193: active_connection_id_limit interacts poorly with Retire Prior To
The `active_connection_id_limit` parameters communicates the maximum number of connection IDs sent in NEW_CONNECTION_ID frames that an endpoint is willing to store. This value is only advisory though, there's nothing stopping an implementation from storing more CIDs than this limit, and the peer is allowed to send more CIDs than this limit:
> An endpoint SHOULD NOT provide more connection IDs than the peer’s limit.

We don't specify any behavior in the case the peer provides more CIDs than this limit, but since retiring a CID asks the peer to provide a new CID, it seems like the reasonable thing to do would be to silently drop the CID on the floor:
> Sending a RETIRE_CONNECTION_ID frame indicates that the connection ID will not be used again and requests that the peer replace it with a new connection ID using a NEW_CONNECTION_ID frame.

This means that a peer cannot fulfill its duty of retiring those discarded connection IDs when receiving a NEW_CONNECTION_ID frame with Retire Prior To set:
>  Upon receipt, the peer MUST retire the corresponding connection IDs and send corresponding RETIRE_CONNECTION_ID frames.

This, however, is not guaranteed, either. A peer might decide to send a RETIRE_CONNECTION_ID frame for a sequence number that it already forgot.

Furthermore, depending on how implementations decide with all these uncertainties, one could end up in a situation where an endpoint only has a single active CID available, but its peer assumes that it is still keeping track of `active_connection_id_limit` CIDs and thus won't provide any new CIDs. In this situation, the endpoint won't be able to initiate or respond to a connection migration any more, leading to a failure of the connection.

### LINKED ISSUE #3197: active_connection_id_limit defaults to 0
The implication, of course, is that 0 => infinite, but we don't actually say that anywhere.  (This part might be editorial; it omits to say what we all know it means.)

1 is also a mostly-invalid value, since you can't avoid exceeding that limit without the peer dropping to zero CIDs in the interim and therefore stalling the connection for an RTT + any recovery time needed.  Also, if the server uses `preferred_address`, it has provided two CIDs before it knows the client's limit.

This is fine, really, so long as the peer SHOULD NOT exceed the limit; if you send (or default to) a stupid value, the peer will ignore you and all is well.  But if we take #3193's suggestion that exceeding the limit be a MUST NOT, the default needs to be infinite and values less than 2 need to be prohibited.

### LINKED ISSUE #3200: Definition of "active connection ID" is misleading (or the name is)
In [section 5.1.1](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.5.1.1), we state that:
> An endpoint SHOULD ensure that its peer has a sufficient number of available and unused connection IDs. Endpoints store received connection IDs for future use and advertise the number of connection IDs they are willing to store with the active_connection_id_limit transport parameter. An endpoint SHOULD NOT provide more connection IDs than the peer’s limit.
> 
> An endpoint SHOULD supply a new connection ID when it receives a packet with a previously unused connection ID or when the peer retires one, unless providing the new connection ID would exceed the peer’s limit.

I think that these are correct. However, we also state:
> Connection IDs that are issued and not retired are considered active; any active connection ID can be used.

I think this is incorrect. This should instead state that "Connection IDs that are **unused and** not retired are considered active, ..."

That said, I think the name is confusing, as it is awkward to argue that the connection ID that is currently used is "active."

Maybe we should rename the parameter to something like "spare_connection_id_limit."

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #3193.
Fixes #3197.
Fixes #3200.

### CODE REVIEW

**Comment by marten-seemann** [comment]
@erickinnear @MikeBishop As discussed in #3197 I added a minimum value of 2 for the TP, since values below this don't make any sense. 
Having an invalid default value seems wrong to me, so I also made 2 the default value for the TP. People might disagree if this a reasonable value.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -999,8 +999,10 @@ retired are considered active; any active connection ID can be used.
 An endpoint SHOULD ensure that its peer has a sufficient number of available and
 unused connection IDs.  Endpoints store received connection IDs for future use
 and advertise the number of connection IDs they are willing to store with the
-active_connection_id_limit transport parameter.  An endpoint SHOULD NOT provide
-more connection IDs than the peer's limit.
+active_connection_id_limit transport parameter.  An endpoint MUST NOT provide
+more connection IDs than the peer's limit.  An endpoint that receives more
+connection IDs than its advertised active_connection_id_limit MUST close the
+connection with an error of type CONNECTION_ID_LIMIT_ERROR.
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: We're requiring other limit checks (e.g. for STREAM_LIMIT_ERROR and for FLOW_CONTROL_ERROR), so I'd prefer to keep this a requirement as well.

I don't think dropping should be regarded as acceptable behavior (although you can't prevent implementations from doing it), since a CID issued by an endpoint consumes state on that endpoint as well.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -999,8 +999,10 @@ retired are considered active; any active connection ID can be used.
 An endpoint SHOULD ensure that its peer has a sufficient number of available and
 unused connection IDs.  Endpoints store received connection IDs for future use
 and advertise the number of connection IDs they are willing to store with the
-active_connection_id_limit transport parameter.  An endpoint SHOULD NOT provide
-more connection IDs than the peer's limit.
+active_connection_id_limit transport parameter.  An endpoint MUST NOT provide
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: This text indicates that the number of "active_connection_id_limit" means the number of "stored" CIDs "for future use." It might be better to address this now. See #3200.

Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: @kazuho I updated the PR such that the active_connection_id_limit now included the handshake and SPA CID, which should make it a fix for #3200 now. PTAL.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -4590,8 +4595,11 @@ preferred_address (0x000d):
 active_connection_id_limit (0x000e):
 
 : The maximum number of connection IDs from the peer that an endpoint is willing
-  to store. This value includes only connection IDs sent in NEW_CONNECTION_ID
-  frames. If this parameter is absent, a default of 0 is assumed.
+  to store. This value includes connection IDs sent in NEW_CONNECTION_ID frames,
+  the connection ID used during the handshake as well as the connection ID
+  provided in the preferred_address transport parameter.
+  If the active_connection_id_limit parameter is absent, a default of 2 is
+  assumed. Values below 2 are invalid.
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: The default and minimal values of 2 look odd, when using a zero-length CID.

Maybe rephrase to:
> Unless a zero-length connection ID is being used, the value of the active_connection_id_limit parameter MUST be no less than 2. If the active_connection_id_limit parameter is absent, a default of 2 is
   assumed.
> 
> When a zero-length connection ID is being used, the active_connection_id_limit parameter MUST NOT be sent.

The rationale behind the last sentence is that it disputable if we consider CIDL=0 as the absence of CID or as the use of zero-length CID. The value of active_connection_id_limit becomes zero or one based on the view you choose. Hence my preference to require absence of the TP when a zero-length CID is used.

Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: Thanks for catching this. I updated the PR.

File: draft-ietf-quic-transport.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: I'm in agreement with @MikeBishop -- I would suggest changing this to a SHOULD error. This is different than the flow control errors, since an endpoint does not have to store all connection IDs that have been received.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -1033,10 +1036,12 @@ longer plans to use that address.
 
 An endpoint can cause its peer to retire connection IDs by sending a
 NEW_CONNECTION_ID frame with an increased Retire Prior To field.  Upon receipt,
-the peer MUST retire the corresponding connection IDs using RETIRE_CONNECTION_ID
-frames.  Failure to retire the connection IDs within approximately one PTO can
-cause packets to be delayed, lost, or cause the original endpoint to send a
-stateless reset in response to a connection ID it can no longer route correctly.
+the peer MUST first retire the corresponding connection IDs using
+NEW_CONNECTION_ID frames and then add the newly provided connection ID to the
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
RETIRE_CONNECTION_ID frames and then add the newly provided connection ID to the
```

File: draft-ietf-quic-transport.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kazuho) [reply draft-ietf-quic-transport.md]: I might suggest having a design discussion in the issue (i.e. #3193) rather than here. There has been certain amount of discussion in that issue, and it seems to me that we have preference for MUST NOT.

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -4590,8 +4595,14 @@ preferred_address (0x000d):
...
-  to store. This value includes only connection IDs sent in NEW_CONNECTION_ID
-  frames. If this parameter is absent, a default of 0 is assumed.
+  to store. This value includes connection IDs sent in NEW_CONNECTION_ID frames,
+  the connection ID used during the handshake as well as the connection ID
+  provided in the preferred_address transport parameter.
+  Unless a zero-length connection ID is being used, the value of the
+  active_connection_id_limit parameter MUST be no less than 2. If this
+  transport parameter is absent, a default of 2 is assumed.
+  When a zero-length connection ID is being used, the active_connection_id_limit
+  parameter MUST NOT be sent.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I think this change is unnecessary and detrimental. Why is this useful?

File: draft-ietf-quic-transport.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (erickinnear) [reply draft-ietf-quic-transport.md]: I'm not sure that I'm against the SHOULD, but I think the idea behind enforcing this was that an endpoint would indeed now need to store every CID it receives, and if it's not willing to store X amount, then it shouldn't have advertised a limit of X.

And even if you *still* didn't store them all, you technically don't need to store them to keep track of how many you've been issued, which you'll probably need in order to cope with getting Retire Prior To anyways.

File: draft-ietf-quic-transport.md (hunk 5 above)
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: Can you elaborate which part you find detrimental?

Comment (kazuho) [reply draft-ietf-quic-transport.md]: @janaiyengar PTAL at https://github.com/quicwg/base-drafts/pull/3201#discussion_r347122022.

Comment (erickinnear) [reply draft-ietf-quic-transport.md]: (Would you rather it was ignored rather than not sent?)

File: draft-ietf-quic-transport.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: I'm ok with this being a MUST. I would suggest that you explicitly add a note saying that an endpoint is expected to store enough information all received CIDs to be able to generate this error. It might be redundant, but that's ok.

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -999,12 +999,15 @@ retired are considered active; any active connection ID can be used.
...
-the new connection ID would exceed the peer's limit.  An endpoint MAY limit the
+active_connection_id_limit transport parameter.  An endpoint MUST NOT provide
+more connection IDs than the peer's limit.  An endpoint that receives more
+connection IDs than its advertised active_connection_id_limit MUST close the
+connection with an error of type CONNECTION_ID_LIMIT_ERROR.
+
+An endpoint SHOULD supply a new connection ID when the peer retires a connection
+ID. If an endpoint provided fewer connection IDs than the peer's
+active_connection_id_limit, it MAY supply a new connection ID when it receives
+a packet with a previously unused connection ID.  An endpoint MAY limit the
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Is this MAY necessary? It's certainly not disallowed. I understand how this could be useful, but it seems like a purely server-side optimization, which doesn't need to be specified. Alternatively, how about something like: "An endpoint can but is not required to send as many connection IDs as the peer can store."

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -4590,8 +4595,14 @@ preferred_address (0x000d):
 active_connection_id_limit (0x000e):
 
 : The maximum number of connection IDs from the peer that an endpoint is willing
-  to store. This value includes only connection IDs sent in NEW_CONNECTION_ID
-  frames. If this parameter is absent, a default of 0 is assumed.
+  to store. This value includes connection IDs sent in NEW_CONNECTION_ID frames,
+  the connection ID used during the handshake as well as the connection ID
+  provided in the preferred_address transport parameter.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Suggested edit: "This value includes the connection ID received during the handshake, that received in the preferred_address transport parameter, and those received in NEW_CONNECTION_ID frames."

File: draft-ietf-quic-transport.md (hunk 5 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: I said detrimental because we used to not encode values on the wire that were illegal ... but we've walked away from that in other places too :-/ I'm ok with this.

File: draft-ietf-quic-transport.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (huitema) [reply draft-ietf-quic-transport.md]: I don't think this shhould be MUST. For every other such action, the paradigm is "the sender MUST NOT do this silly stuff, the responder MAY close with connection error"

[the remark this answers is not in the document]
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: @huitema That's not correct:
> Endpoints MUST NOT exceed the limit set by their peer. An endpoint that receives a frame with a stream ID exceeding the limit it has sent MUST treat this as a connection error of type STREAM_LIMIT_ERROR 

and 
>  An endpoint MUST terminate a connection with a FLOW_CONTROL_ERROR error if it receives more data than the maximum data value that it has sent

File: draft-ietf-quic-transport.md (hunk 6 above)
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: This sentence goes back to @kazuho's issue https://github.com/quicwg/base-drafts/issues/2436.

File: draft-ietf-quic-transport.md (hunk 7 above)
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: Applied. Thank you.

**Comment by martinthomson** [comment]
I'm seeing some people asking for changes, but I am happy with Marten's responses to those.  I'm going to mark the associated issues as `proposal-ready`.

File: draft-ietf-quic-transport.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (huitema) [reply draft-ietf-quic-transport.md]: @marten-seemann The examples that you quote are cases where there is no reasonable way to continue the connection. I don't think that excess number of NEW CONNECTION ID is in the same class -- the receiving peer could also just drop them without retiring them.

[the remark this answers is not in the document]
Comment (marten-seemann) [reply draft-ietf-quic-transport.md]: @huitema That doesn’t work well with Retire Prior To, since you still MUST retire that connection ID. So you need to track at least the sequence number of the CID.
It seems like you’re trying to build a way for a peer to (silently) revoke the limit it announced in its transport parameters. Building such a mechanism is possible (we’d need something like a MAX_CID frame), but we decided to not do this, mostly because we assumed that the number of requested connection IDs will be small (O(10)). If that argument still holds, I don’t think it’s too much to ask for to keep track of O(10) CID sequence numbers.
